In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Permit Data

In [2]:
sf_permits = pd.read_csv('Building_Permits.csv')
sf_permits.head()

C:\Users\Kendra\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2728: DtypeWarning: Columns (22,32) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           05/06/2015  0326  023            140                  NaN   
1           04/19/2016  0306  007            440                  NaN   
2           05/27/2016  0595  203           1647                  NaN   
3           11/07/2016  0156  011           1230                  NaN   
4           11/28/2016  0342  001            950                  NaN   

  Street Name Street Suffix      ...        Existing Construction Type  \
0       Ellis            St      ...                               3.0   
1       Geary            St      ...                               3.0   
2     Pacific            Av      ...                               1.0   
3     Pacific            Av      ...                               5.0   
4      Market            St      ...                               3.0   

  Existing Construction Type Description Proposed Construction Type  \
0                          constr type 3                        NaN   
1                          constr type 3                        NaN   
2                          constr type 1                        1.0   
3                         wood frame (5)                        5.0   
4                          constr type 3                        NaN   

  Proposed Construction Type Description Site Permit Supervisor District  \
0                                    NaN         NaN                 3.0   
1                                    NaN         NaN                 3.0   
2                          constr type 1         NaN                 3.0   
3                         wood frame (5)         NaN                 3.0   
4                                    NaN         NaN                 6.0   

  Neighborhoods - Analysis Boundaries  Zipcode  \
0                          Tenderloin  94102.0   
1                          Tenderloin  94102.0   
2                        Russian Hill  94109.0   
3                            Nob Hill  94109.0   
4                          Tenderloin  94102.0   

                                    Location      Record ID  
0  (37.785719256680785, -122.40852313194863)  1380611233945  
1   (37.78733980600732, -122.41063199757738)  1420164406718  
2    (37.7946573324287, -122.42232562979227)  1424856504716  
3   (37.79595867909168, -122.41557405519474)  1443574295566  
4   (37.78315261897309, -122.40950883997789)   144548169992  

[5 rows x 43 columns]

How much data is missing?

In [3]:
missing_perm_counts = sf_permits.isnull().sum()
missing_perm_counts.head(10)

Permit Number                  0
Permit Type                    0
Permit Type Definition         0
Permit Creation Date           0
Block                          0
Lot                            0
Street Number                  0
Street Number Suffix      196684
Street Name                    0
Street Suffix               2768
dtype: int64

Express as percent of total # entries:

In [4]:
total_perm_rows = len(sf_permits)
missing_perm_df = pd.DataFrame({'Missing Data Counts':missing_perm_counts})
missing_perm_df['Missing Data %'] = missing_perm_df['Missing Data Counts'] / total_perm_rows *100
missing_perm_df.head(20)

Missing Data Counts  Missing Data %
Permit Number                                       0        0.000000
Permit Type                                         0        0.000000
Permit Type Definition                              0        0.000000
Permit Creation Date                                0        0.000000
Block                                               0        0.000000
Lot                                                 0        0.000000
Street Number                                       0        0.000000
Street Number Suffix                           196684       98.885872
Street Name                                         0        0.000000
Street Suffix                                    2768        1.391654
Unit                                           169421       85.178984
Unit Suffix                                    196939       99.014077
Description                                       290        0.145802
Current Status                                      0        0.000000
Current Status Date                                 0        0.000000
Filed Date                                          0        0.000000
Issued Date                                     14940        7.511312
Completed Date                                 101709       51.135747
First Construction Document Date                14946        7.514329
Structural Notification                        191978       96.519859

### Which data doesn't exist, and which simply wasn't recorded?

Look at 'Street Number Suffix':

In [5]:
sf_permits[sf_permits['Street Number Suffix'].notnull()][0:5]

Permit Number  Permit Type       Permit Type Definition  \
92   201408254591            8       otc alterations permit   
146       M803987            8       otc alterations permit   
164  201707141931            2  new construction wood frame   
200       M816647            8       otc alterations permit   
273       M829468            8       otc alterations permit   

    Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
92            08/25/2014  3509  029             63                    A   
146           07/03/2017  0821  073            843                    C   
164           07/14/2017  1193  048           2150                    A   
200           08/04/2017  0821  073            843                    C   
273           09/08/2017  3702  390             33                    A   

    Street Name Street Suffix      ...        Existing Construction Type  \
92        Grace            St      ...                               5.0   
146       Hayes            St      ...                               NaN   
164       Hayes            St      ...                               NaN   
200       Hayes            St      ...                               NaN   
273        08th            St      ...                               NaN   

    Existing Construction Type Description Proposed Construction Type  \
92                          wood frame (5)                        5.0   
146                                    NaN                        NaN   
164                                    NaN                        5.0   
200                                    NaN                        NaN   
273                                    NaN                        NaN   

    Proposed Construction Type Description Site Permit Supervisor District  \
92                          wood frame (5)         NaN                 6.0   
146                                    NaN         NaN                 5.0   
164                         wood frame (5)           Y                 5.0   
200                                    NaN         NaN                 5.0   
273                                    NaN         NaN                 6.0   

    Neighborhoods - Analysis Boundaries  Zipcode  \
92                      South of Market  94103.0   
146                        Hayes Valley  94117.0   
164                   Lone Mountain/USF  94117.0   
200                        Hayes Valley  94117.0   
273                     South of Market  94103.0   

                                      Location      Record ID  
92   (37.774944547030586, -122.41443674356925)  1353299454075  
146  (37.775671880692904, -122.43022136710104)  1469150425694  
164   (37.77347316353236, -122.45200535060522)   147059391960  
200  (37.775671880692904, -122.43022136710104)  1473476425694  
273     (37.77850219390184, -122.413426065671)  1478111509834  

[5 rows x 43 columns]

Appears to be an Apartment letter. Do not expect every entry to have one.

What about Zipcode?

In [6]:
missing_perm_df.loc['Zipcode']

Missing Data Counts    1716.000000
Missing Data %            0.862745
Name: Zipcode, dtype: float64

In [7]:
sf_permits[sf_permits['Zipcode'].isnull()][0:5]

Permit Number  Permit Type  Permit Type Definition Permit Creation Date  \
452       M866367            8  otc alterations permit           12/13/2017   
464       M845627            8  otc alterations permit           10/18/2017   
577       M364427            8  otc alterations permit           01/02/2013   
731  201301047237            8  otc alterations permit           01/04/2013   
864  201301077331            8  otc alterations permit           01/07/2013   

     Block  Lot  Street Number Street Number Suffix Street Name Street Suffix  \
452   0552  017           2550                  NaN       Gough            St   
464  5286A  021           2241                  NaN     Jerrold            Av   
577   0779  028           1235                  NaN  Mcallister            St   
731   6515  017           1387                  NaN    Valencia            St   
864   3706  095             55                  NaN        04th            St   

         ...        Existing Construction Type  \
452      ...                               NaN   
464      ...                               NaN   
577      ...                               NaN   
731      ...                               5.0   
864      ...                               1.0   

    Existing Construction Type Description Proposed Construction Type  \
452                                    NaN                        NaN   
464                                    NaN                        NaN   
577                                    NaN                        NaN   
731                         wood frame (5)                        5.0   
864                          constr type 1                        1.0   

    Proposed Construction Type Description Site Permit Supervisor District  \
452                                    NaN         NaN                 NaN   
464                                    NaN         NaN                 NaN   
577                                    NaN         NaN                 NaN   
731                         wood frame (5)         NaN                 NaN   
864                          constr type 1         NaN                 NaN   

    Neighborhoods - Analysis Boundaries Zipcode Location      Record ID  
452                                 NaN     NaN      NaN  1490471235302  
464                                 NaN     NaN      NaN  1490955510104  
577                                 NaN     NaN      NaN   129220681682  
731                                 NaN     NaN      NaN  1292365379064  
864                                 NaN     NaN      NaN  1292500249105  

[5 rows x 43 columns]

Missing Zipcode entries are likely so because they weren't recorded.

Maybe we can guess what they are based on Block ID?

In [8]:
sf_permits[sf_permits['Block'] == '0552'][0:5]

Permit Number  Permit Type  Permit Type Definition Permit Creation Date  \
452        M866367            8  otc alterations permit           12/13/2017   
760   201301047258            8  otc alterations permit           01/04/2013   
761   201301047258            8  otc alterations permit           01/04/2013   
1218       M365467            8  otc alterations permit           01/09/2013   
6329  201303081803            8  otc alterations permit           03/08/2013   

     Block  Lot  Street Number Street Number Suffix Street Name Street Suffix  \
452   0552  017           2550                  NaN       Gough            St   
760   0552  006           2425                  NaN    Franklin            St   
761   0552  006           2429                  NaN    Franklin            St   
1218  0552  023           1629                  NaN       Green            St   
6329  0552  010           1730                  NaN     Vallejo            St   

          ...        Existing Construction Type  \
452       ...                               NaN   
760       ...                               5.0   
761       ...                               5.0   
1218      ...                               NaN   
6329      ...                               5.0   

     Existing Construction Type Description Proposed Construction Type  \
452                                     NaN                        NaN   
760                          wood frame (5)                        5.0   
761                          wood frame (5)                        5.0   
1218                                    NaN                        NaN   
6329                         wood frame (5)                        5.0   

     Proposed Construction Type Description Site Permit Supervisor District  \
452                                     NaN         NaN                 NaN   
760                          wood frame (5)         NaN                 2.0   
761                          wood frame (5)         NaN                 2.0   
1218                                    NaN         NaN                 2.0   
6329                         wood frame (5)         NaN                 2.0   

     Neighborhoods - Analysis Boundaries  Zipcode  \
452                                  NaN      NaN   
760                               Marina  94123.0   
761                               Marina  94123.0   
1218                              Marina  94123.0   
6329                              Marina  94123.0   

                                       Location      Record ID  
452                                         NaN  1490471235302  
760   (37.796839087299894, -122.42563588126309)   129239674044  
761   (37.796839087299894, -122.42563588126309)  1292397420161  
1218    (37.7970555468036, -122.42595789597318)  1292865466352  
6329  (37.796682532721235, -122.42588184378776)  1298222235298  

[5 rows x 43 columns]

In [9]:
sf_permits[sf_permits['Block'] == '1040']['Zipcode'].unique()

array([94118.,    nan])

Yup, all permits with a given Block ID have the same Zipcode.

#### Create dictionary of block-zipcode pairs:

In [10]:
block_zip_dict = {}
for b in sf_permits[sf_permits['Zipcode'].isnull()]['Block'].unique():
        l = sf_permits[sf_permits['Block'] == b]['Zipcode'].unique()
        if len(l) > 1:
            z = l[np.isnan(l) == False][0]
        else:
            z = None
        block_zip_dict[b] = z

    

In [11]:
block_zip_dict

{'0000': None,
 '0004': None,
 '0005': None,
 '0025': 94109.0,
 '0027': 94109.0,
 '0028': 94133.0,
 '0039': 94133.0,
 '0041': 94133.0,
 '0044': 94133.0,
 '0051': 94133.0,
 '0052': 94133.0,
 '0054': 94133.0,
 '0060': 94111.0,
 '0061C': None,
 '0062': 94133.0,
 '0066': 94133.0,
 '0067': 94133.0,
 '0069': 94109.0,
 '0072': 94133.0,
 '0074': 94133.0,
 '0078': 94133.0,
 '0085': 94133.0,
 '0087': 94133.0,
 '0090': 94133.0,
 '0093': 94133.0,
 '0095': 94109.0,
 '0097': 94109.0,
 '00979': 94115.0,
 '0098': 94133.0,
 '0101': 94133.0,
 '0103': 94133.0,
 '0106': 94133.0,
 '0113': 94111.0,
 '0114': 94133.0,
 '0115': 94133.0,
 '0120': 94133.0,
 '0122': 94109.0,
 '0123': 94109.0,
 '0134': 94133.0,
 '0141': 94111.0,
 '0148': 94133.0,
 '0150': 94133.0,
 '0151': 94109.0,
 '0153': 94109.0,
 '0159': 94133.0,
 '0160': 94133.0,
 '0164': 94133.0,
 '0168': 94111.0,
 '0176': 94133.0,
 '0188': 94109.0,
 '0190': 94108.0,
 '0206': 94111.0,
 '0213': 94108.0,
 '0215': 94109.0,
 '0216': 94109.0,
 '0218': 94109.0,
 '

Apply to df:

In [12]:
sf_permits.loc[sf_permits['Zipcode'].isnull(), 'Zipcode'] = sf_permits.loc[sf_permits['Zipcode'].isnull(), 'Block'].map(block_zip_dict)

Check that a previously 'nan' entry (index #452) is now populated;

In [13]:
sf_permits.loc[450:455,['Block','Zipcode']]

Block  Zipcode
450  0467A  94123.0
451   0629  94115.0
452   0552  94123.0
453   0559  94123.0
454   3953  94107.0
455   3540  94114.0

#### BOOM.